src/schedule/zephyr_dp_schedule_native.c

```diff
@@ -0,0 +1,572 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright(c) 2025 Intel Corporation. All rights reserved.
+ *
+ * Author: Marcin Szkudlinski
+ */
+
+#include <rtos/kernel.h>
+#include <rtos/task.h>
+
+#include <sof/audio/module_adapter/module/generic.h>
+#include <sof/common.h>
+#include <sof/list.h>
+#include <sof/llext_manager.h>
+#include <sof/schedule/dp_schedule.h>
+#include <sof/schedule/ll_schedule_domain.h>
+#include <sof/trace/trace.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/slist.h>
+#include <zephyr/app_memory/mem_domain.h>
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "zephyr_dp_schedule.h"
+
+LOG_MODULE_DECLARE(dp_schedule, CONFIG_SOF_LOG_LEVEL);
+extern struct tr_ctx dp_tr;
+
+/* FIXME: memory domains cannot be removed. Need an alternative solution */
+#ifdef DYNAMIC_MEM_DOMAIN
+extern sys_slist_t xtensa_domain_list;
+#endif
+
+#if CONFIG_USERSPACE
+static struct k_mem_domain dp_mdom;
+#endif
+
+/* Synchronization semaphore for the scheduler thread to wait for DP startup */
+#define DP_SYNC_INIT(i, _)	Z_SEM_INITIALIZER(dp_sync[i], 0, 1)
+#define DP_SYNC_INIT_LIST	LISTIFY(CONFIG_CORE_COUNT, DP_SYNC_INIT, (,))
+static STRUCT_SECTION_ITERABLE_ARRAY(k_sem, dp_sync, CONFIG_CORE_COUNT) = { DP_SYNC_INIT_LIST };
+
+/* TODO: make this a shared kernel->module buffer for IPC parameters */
+static uint8_t ipc_buf[4096] __aligned(4096);
+
+struct ipc4_flat {
+	unsigned int cmd;
+	int ret;
+	union {
+		struct {
+			struct ipc4_module_bind_unbind bu;
+			enum bind_type type;
+		} bind;
+		struct {
+			unsigned int trigger_cmd;
+			enum ipc4_pipeline_state state;
+			int n_sources;
+			int n_sinks;
+			void *source_sink[];
+		} pipeline_state;
+	};
+};
+
+/* Pack IPC input data */
+static int ipc_thread_flatten(unsigned int cmd, union scheduler_dp_thread_ipc_param *param,
```

Comment:
Do we really need to pack params into `struct scheduler_dp_thread_ipc_param` only to repack it into `struct ipc4_flat` later? This looks copy-heavy.

Reply:
the first "packing doesn't do a lot of copying, it collects pointers. Whereas the second one has to handle data indeed